libp2p/switch.nim

```diff
@@ -50,12 +50,19 @@ const
 type
     NoPubSubException* = object of CatchableError
 
-    Lifecycle* {.pure.} = enum
-      Connected,
-      Upgraded,
-      Disconnected
+    PeerEventKind* {.pure.} = enum
+      Upgraded, # Peer connected and went through secure upgrade
+      Disconnected # Peer disconnected after upgrade
 
-    Hook* = proc(peer: PeerInfo, cycle: Lifecycle): Future[void] {.gcsafe.}
+    PeerEvent* = object
```

Comment:
Are this peer events or connection events? We should not confuse the two, it's a subtle but fundamental distinction. For context, the peer will have many connections (ie sockets, for some transports) associated with it and it can be considered as disconnected only after the last connection ceased to exist. It isn't clear from the code if that is the case ATM, some clarification would be appropriate. We can/should also have connection events for that reason, which is what the hooks where trying to accomplish.

The Peer lifetime is also crucial to have for an upcoming refactor of the pubsub subsystem, so if this is exactly that then it's a welcome addition, if not, then the renaming of the connection hooks into peer lifetime is at the very least confusing and should be reconsidered.

Reply:
well, they're events related to peers - thanks to the object and kind field, it's easy to add related events as well. updating the docs a bit - this is enough to solve what nbc needs for now.

Reply:
yes, it might be enough for NBC, but they interfere with changes I have in the pipeline for peer lifetime which I rely on heavily in a refactor of pubsub that I've started just recently. Can we please rename this to `connection events`, ie, connection `Upgraded` and connection `Disconnected` otherwise this is plain confusing, specially in lew of the upcoming peer lifetime changes.